src/query/builder.cr

```diff
@@ -0,0 +1,109 @@
+# Data structure which will allow chaining of query components,
+# nesting of boolean logic, etc.
+#
+# Should return self, or another instance of Builder wherever
+# chaining should be possible.
+#
+# Current query syntax:
+# - where(field: value) => "WHERE field = 'value'"
+#
+# Hopefully soon:
+# - Model.where(field: value).not( Model.where(field2: value2) )
+# or
+# - Model.where(field: value).not { where(field2: value2) }
+#
+# - Model.where(field: value).or( Model.where(field3: value3) )
+# or
+# - Model.where(field: value).or { whehre(field3: value3) }
+class Query::Builder(T)
+  alias FieldName = String
+  alias FieldData = DB::Any
+
+  def initialize(@boolean_operator = :and)
+    @fields = {} of FieldName => FieldData
+    @order  = [] of NamedTuple(field: String, direction: String)
+  end
+
+  def compile
+    Compiled(T).new self
+  end
+
+  def runner
+    Runner(T).new compile
+  end
+
+  def where(**matches)
+    matches.each do |field, data|
+      @fields[field.to_s] = data
+    end
+
+    self
+  end
+
+  def order(field : Symbol)
+    @order << { field: field, direction: :ascending }
+
+    self
+  end
+
+  def order(**dsl)
+    dsl.each do |field, dsl_direction|
+      direction = "ASC"
+
+      if dsl_direction == "desc" || dsl_direction == :desc
+        direction = "DESC"
+      end
+
+      @order << { field: field.to_s, direction: direction }
+    end
+
+    self
+  end
+
+  def _build_order
```

Comment:
Are these underscore methods meant to indicate internal use? If so, make them private.

Reply:
Yes and no. They can't be private because they're called by the `Query::Compiled` class. I think this is evidence of an architectural mis-step in this design, but I can't quite see what the correction would be yet.

Reply:
This doesn't really change things, but I like it a little more:
```crystal
def order
  default_order if @order.none?
  build_order
end

private def build_order
  @order.map { |expression| "#{expression[:field]} #{expression[:direction]}" }.join(", ")
end
```

Reply:
`protected` might be what you are looking for

Reply:
Actually, technically, `_build_order` is returning an order clause, right? so:

`def order` - `def order_clause : String`
`private def build_order` -> `private def build_order_clause : String`

Reply:
@eliasjpr - There isn't any inheritance, an instance of this class is used in class he mentioned.

Some of these classes could potentially be made into module mixins, just going off names:
`Query` - noun, class
A query you can build, compile, and run - those could be modules that the `Query` mixes in